### ISSUE TITLE
Getting DDP support models that has part of the parameters/layers unused

### ISSUE THREAD

**Issue by teng-li** [pytorch/pytorch#13273]

**Comment by phi-go** [comment]
Hi, I'm facing this problem as I want to train a model that depending on the input might not use some parameters/layers. Maybe the following sample code will help. As expected without the changes in the pull request this code fails.

```
TypeError: _queue_reduction(): incompatible function arguments. The following argument types are supported:
    1. (process_group: torch.distributed.ProcessGroup, grads_batch: List[List[at::Tensor]], devices: List[int]) -> Tuple[torch.distributed.Work, at::Tensor]

Invoked with: <torch.distributed.ProcessGroupNCCL object at 0x7fa6dfd804c8>, [[None, None, tensor([[-6.3432e+11, -2.8616e+12, -1.3041e+12],
        [-3.8208e+12, -1.7236e+13, -7.8549e+12],
        [-3.4752e+12, -1.5678e+13, -7.1446e+12],
        [-2.8048e+12, -1.2653e+13, -5.7663e+12]], device='cuda:0'), tensor([ -387039.5000, -2331274.5000, -2120455.7500, -1711401.0000],
       device='cuda:0')]], [0]
```

However even with them applied:

```
terminate called after throwing an instance of 'gloo::EnforceNotMet'
  what():  [enforce fail at /opt/conda/conda-bld/pytorch-nightly_1541790704598/work/third_party/gloo/gloo/transport/tcp/pair.cc:456] op.preamble.length <= op.nbytes. 20 vs 16
```

Sample code:

```python
import os
import time

import torch
import torch.distributed as dist
import torch.nn.functional as F
from torch.multiprocessing import Process
from torch.utils.data import DataLoader

from torch.nn.parallel import DistributedDataParallel
from torch.utils.data.distributed import DistributedSampler



class CustomDataset(torch.utils.data.Dataset):
    def __init__(self, data_list):
        self.data_list = data_list

    def __len__(self):
        return len(self.data_list)

    def __getitem__(self, item):
        return self.data_list[item]


class Model(torch.nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.some_used_lin = torch.nn.Linear(1, 2)
        self.lin = torch.nn.Linear(3, 4)

    def forward(self, input):
        input = input.squeeze()
        if input.size(0) == 3:
            return self.lin(input)
        else:
            return torch.cat((self.lin(input[:3]), self.some_used_lin(input[3].unsqueeze(0))))


def get_dataset():
    data_list = [{"x": torch.FloatTensor(3 + (i % 2)).random_(), "y": torch.FloatTensor(4 + (i % 2)*2).random_()}
                 for i in range(10)]
    return CustomDataset(data_list)


def train(rank, size, device_name):
    dataset = get_dataset()
    distributed_sampler = DistributedSampler(dataset, size, rank)
    train_loader = DataLoader(dataset, pin_memory=True, sampler=distributed_sampler)
    device = torch.device(device_name)
    model = Model()
    model.to(device)
    model = DistributedDataParallel(model, device_ids=[device])
    optimizer = torch.optim.Adam(model.parameters())

    for epoch in range(2):
        model.train()
        for batch_num, batch in enumerate(train_loader):
            print("rank", rank, batch['x'].shape)
            optimizer.zero_grad()
            pred = model(batch['x'].to(device))
            exp = batch['y'].to(device).squeeze()
            loss = F.binary_cross_entropy_with_logits(exp, pred)
            loss.backward()
            del loss

            optimizer.step()


def init_processes(rank, size, fn, ddp_args, device):
    """ Initialize the distributed environment. """
    os.environ['MASTER_ADDR'] = ddp_args['master_addr']
    os.environ['MASTER_PORT'] = ddp_args['master_port']
    os.environ['WORLD_SIZE'] = str(size)
    os.environ['RANK'] = str(rank)
    dist.init_process_group(ddp_args['backend'], rank=rank, world_size=size)
    fn(rank, size, device)
    print("done")


def main():
    ddp_args = {"master_addr": "127.0.0.1", "master_port": "29500", "backend": "gloo"}
    devices = ["cuda:0", "cuda:0"]
    size = len(devices)
    processes = []
    for rank, device in enumerate(devices):
        p = Process(target=init_processes, args=(rank, size, train, ddp_args, device))
        p.start()
        processes.append(p)

    while not any(p.exitcode is not None for p in processes):  # all worker processes have finished
        time.sleep(0.2)

    for p in processes:
        print("join")
        p.join()
        print("join done")
    print("all done")


if __name__ == "__main__":
    main()
```
Env:

```
PyTorch version: 1.0.0.dev20181109
Is debug build: No
CUDA used to build PyTorch: 9.2.148

OS: Arch Linux
GCC version: (GCC) 8.2.1 20180831
CMake version: version 3.12.4

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 410.66
```

**Comment by acgtyrant** [comment]
I met the same error as phi-go when I upgrade torch from 0.4.1 to 1.0.0

**Comment by teng-li** [comment]
DDP currently doesn't support this, the old DDP in 0.4.1 doesn't support this either but it just happened to run (not necessarily work) without throwing error.

For now, can you guys make sure all parameters in the models are used. I will see if I can get DDP support this soon.

Thanks for your patience

**Comment by yoershine** [comment]
I met the same error using distributed_c10d when freezing part of module parameters.

**Comment by PetrochukM** [comment]
@teng-li Hi There! This is an oddball question but during testing a neural network, it'd be useful to check if all the parameters are used. Any idea of how to unit test that?

**Comment by mcarilli** [comment]
In the meantime you can try our distributed module wrapper, `apex.parallel.DistributedDataParallel`
[source](https://github.com/NVIDIA/apex/tree/master/apex/parallel) 
[docs](https://nvidia.github.io/apex/parallel.html) 
[simple example](https://github.com/NVIDIA/apex/tree/master/examples/FP16_Optimizer_simple/distributed_apex) (ignore the stuff with FP16_Optimizer) 
[Imagenet examples](https://github.com/NVIDIA/apex/tree/master/examples/imagenet) (again, ignore the mixed precision/amp/FP16_Optimizer logic.  That's unrelated.  Just look at where DDP is imported and constructed.)

The API is almost identical to `torch.nn.parallel.DistributedDataParallel` (the big difference is that you do not need to supply the device_ids argument).

With `delay_allreduce=True`, Apex DDP should handle any of the above use cases (freezing, or None gradients).
With `delay_allreduce=False` (aggressively overlap comms) Apex DDP should be able to handle freezing.  It should also be able to handle None gradients **as long as the set of params that do not receive gradients remains the same from iteration to iteration**.

A limitation of apex DDP is that it requires you to use only one device per process.

If the Apex DDP logic turns out to support your use cases, we will know that it's worth upstreaming.  Once that's accomplished it should be easy to switch back to `torch.nn.parallel.DistributedDataParallel` because the APIs are quite similar.

**Comment by CoinCheung** [comment]
Have you solved this problem? I met same problem when I train on single node with multi-gpus.

**Comment by xllau** [comment]
> In the meantime you can try our distributed module wrapper, `apex.parallel.DistributedDataParallel`.
> With `delay_allreduce=True`, Apex DDP should handle any of the above use cases (freezing, or None gradients).
> With `delay_allreduce=False` (aggressively overlap comms) Apex DDP should be able to handle freezing. It should also be able to handle None gradients **as long as the set of params that do not receive gradients remains the same from iteration to iteration**.
> 
> [source](https://github.com/NVIDIA/apex/tree/master/apex/parallel)
> [docs](https://nvidia.github.io/apex/parallel.html)
> [simple example](https://github.com/NVIDIA/apex/tree/master/examples/FP16_Optimizer_simple/distributed_apex) (ignore the stuff with FP16_Optimizer)
> [Imagenet examples](https://github.com/NVIDIA/apex/tree/master/examples/imagenet) (again, ignore the mixed precision/amp/FP16_Optimizer logic. That's unrelated. Just look at where DDP is imported and constructed.)
> 
> The API is almost identical to `torch.nn.parallel.DistributedDataParallel` (the big difference is that you do not need to supply the device_ids argument).
> 
> A limitation of apex DDP is that it requires you to use only one device per process.
> 
> If the Apex DDP logic turns out to support your use cases, we will know that it's worth upstreaming. Once that's accomplished it should be easy to switch back to `torch.nn.parallel.DistributedDataParallel` because the APIs are quite similar.

I have met the problem of training error with more than 1 GPU while with 1 GPU is OK. I have replace the distribution part with your recommendation but the problem still happens.

**Comment by geduo15** [comment]
> In the meantime you can try our distributed module wrapper, `apex.parallel.DistributedDataParallel`.
> With `delay_allreduce=True`, Apex DDP should handle any of the above use cases (freezing, or None gradients).
> With `delay_allreduce=False` (aggressively overlap comms) Apex DDP should be able to handle freezing. It should also be able to handle None gradients **as long as the set of params that do not receive gradients remains the same from iteration to iteration**.
> 
> [source](https://github.com/NVIDIA/apex/tree/master/apex/parallel)
> [docs](https://nvidia.github.io/apex/parallel.html)
> [simple example](https://github.com/NVIDIA/apex/tree/master/examples/FP16_Optimizer_simple/distributed_apex) (ignore the stuff with FP16_Optimizer)
> [Imagenet examples](https://github.com/NVIDIA/apex/tree/master/examples/imagenet) (again, ignore the mixed precision/amp/FP16_Optimizer logic. That's unrelated. Just look at where DDP is imported and constructed.)
> 
> The API is almost identical to `torch.nn.parallel.DistributedDataParallel` (the big difference is that you do not need to supply the device_ids argument).
> 
> A limitation of apex DDP is that it requires you to use only one device per process.
> 
> If the Apex DDP logic turns out to support your use cases, we will know that it's worth upstreaming. Once that's accomplished it should be easy to switch back to `torch.nn.parallel.DistributedDataParallel` because the APIs are quite similar.


I met the error when using the torch.nn.Adaptive Log SoftmaxWithLoss in our model, but when I change torch.nn.parallel.DistributedDataParallel into apex ,the torch.nn.Adaptive Log SoftmaxWithLoss still doesn't work(torch.nn.Adaptive Log SoftmaxWithLoss needs part of  parameters be updated but not all parameters). My runtime environment: python3.6, cudnn 9, 8 GPUs

**Comment by denru01** [comment]
Hi all,

I encountered the same issue. Is there any update or workaround?

**Comment by soumith** [comment]
@denru01 #18251 aims to fix this in pytorch core as well

**Comment by pietern** [comment]
This will be solved once #18953 is merged.

It fixes this issue by searching the autograd graph of the output variable (return value from `forward`) and preemptively marking the gradient of every model parameter that isn't part of that graph as ready.

Please give it a try and see if it works for you, and comment here or create a new issue if it doesn't.

**Comment by yaohuaxin** [comment]
Did it upstreamed? @mcarilli

**Comment by sermakarevich** [comment]
it looks like the issue remains even in `torch==1.3`
I need to filter out some part of the model output and this causes the error
```
TypeError: _queue_reduction(): incompatible function arguments. The following argument types are supported:
    1. (process_group: torch.distributed.ProcessGroup, grads_batch: List[List[at::Tensor]], devices: List[int]) -> Tuple[torch.distributed.Work, at::Tensor]
```

any updates ?

**Comment by pietern** [comment]
@sermakarevich An error calling `_queue_reduction` means you're not using `DistributedDataParallel` from PyTorch 1.2 or 1.3. This sounds like either an issue with Apex, or a misconfiguration. Please open a new issue if this is indeed a problem.

**Comment by williamFalcon** [comment]
Any updates on this? Currently having this issue with pt 1.3. Need to calculate a parametrized softmax using the output of DP.

```python
outs = DP.fwd(x)

loss = softmax(outs, proj_params)
```

**Comment by pietern** [comment]
@williamFalcon Can you open a new issue for this?
Please provide an exact description of what you're trying to do and why it's not working as expected.

**Comment by CANEAT** [comment]
> In the meantime you can try our distributed module wrapper, `apex.parallel.DistributedDataParallel`.
> With `delay_allreduce=True`, Apex DDP should handle any of the above use cases (freezing, or None gradients).
> With `delay_allreduce=False` (aggressively overlap comms) Apex DDP should be able to handle freezing. It should also be able to handle None gradients **as long as the set of params that do not receive gradients remains the same from iteration to iteration**.
> 
> [source](https://github.com/NVIDIA/apex/tree/master/apex/parallel)
> [docs](https://nvidia.github.io/apex/parallel.html)
> [simple example](https://github.com/NVIDIA/apex/tree/master/examples/FP16_Optimizer_simple/distributed_apex) (ignore the stuff with FP16_Optimizer)
> [Imagenet examples](https://github.com/NVIDIA/apex/tree/master/examples/imagenet) (again, ignore the mixed precision/amp/FP16_Optimizer logic. That's unrelated. Just look at where DDP is imported and constructed.)
> 
> The API is almost identical to `torch.nn.parallel.DistributedDataParallel` (the big difference is that you do not need to supply the device_ids argument).
> 
> A limitation of apex DDP is that it requires you to use only one device per process.
> 
> If the Apex DDP logic turns out to support your use cases, we will know that it's worth upstreaming. Once that's accomplished it should be easy to switch back to `torch.nn.parallel.DistributedDataParallel` because the APIs are quite similar.

I have same problem when training error with more than 1 GPU while with 1 GPU is OK. it work for me,Thanks.

**Comment by edgarschnfld** [comment]
> In the meantime you can try our distributed module wrapper, `apex.parallel.DistributedDataParallel`.
> With `delay_allreduce=True`, Apex DDP should handle any of the above use cases (freezing, or None gradients).
> With `delay_allreduce=False` (aggressively overlap comms) Apex DDP should be able to handle freezing. It should also be able to handle None gradients **as long as the set of params that do not receive gradients remains the same from iteration to iteration**.
> 
> [source](https://github.com/NVIDIA/apex/tree/master/apex/parallel)
> [docs](https://nvidia.github.io/apex/parallel.html)
> [simple example](https://github.com/NVIDIA/apex/tree/master/examples/FP16_Optimizer_simple/distributed_apex) (ignore the stuff with FP16_Optimizer)
> [Imagenet examples](https://github.com/NVIDIA/apex/tree/master/examples/imagenet) (again, ignore the mixed precision/amp/FP16_Optimizer logic. That's unrelated. Just look at where DDP is imported and constructed.)
> 
> The API is almost identical to `torch.nn.parallel.DistributedDataParallel` (the big difference is that you do not need to supply the device_ids argument).
> 
> A limitation of apex DDP is that it requires you to use only one device per process.
> 
> If the Apex DDP logic turns out to support your use cases, we will know that it's worth upstreaming. Once that's accomplished it should be easy to switch back to `torch.nn.parallel.DistributedDataParallel` because the APIs are quite similar.

This solved it for me. All I had to do was to install apex and replace ```torch.nn.parallel.DistributedDataParallel``` with ```apex.parallel.DistributedDataParallel```, replace the original arguments with ```delay_allreduce = True```.

**Comment by 601222543** [comment]
Have you solved this problem? I met same problem when I train on single node with multi-gpus.

**Comment by richardbaihe** [comment]
> 

Hi @geduo15 , did you find a way to fix this? I am also using adaptive-softmax, which means not all output layer weights of softmax will be used and updated. Currently, data_parallel works for me, but DDP and apexDDP don't work.

**Comment by geduo15** [comment]
I rembered that the problem has been solved in the last year, how about
updated the version of pytorch. In fact, I have successfully trained a
model.




He Bai ***@***.***> 于2021年12月15日周三 05:22写道：
